Spark: Add a test to check if the bloom filters are added to the parquet files

hussein-awala opened the pull request:
related: #9898

This PR refactors `TestSparkReaderWithBloomFilter` tests by replacing avro writer with spark write and makes tests of different versions of spark consistent. It also adds a test new test to check if the bloom filters are correctly added to the parquet files.

amogh-jahagirdar commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderWithBloomFilter.java:
```diff
@@ -367,11 +374,28 @@ public void testReadWithFilter() {
   }
+
+  @TestTemplate
+  public void testBloomCreation() throws IOException {
+    org.apache.hadoop.fs.Path path = new org.apache.hadoop.fs.Path(temp.toString());
+    ParquetMetadata parquetMetadata = ParquetFileReader.readFooter(new Configuration(), path);
+    for(int i = 0; i < 11; i++) {
+      if (useBloomFilter)
+      {
+        assertThat(parquetMetadata.getBlocks().get(0).getColumns().get(0).getBloomFilterOffset()).isNotEqualTo(-1L);
+      }
+      else
+      {
+          assertThat(parquetMetadata.getBlocks().get(0).getColumns().get(0).getBloomFilterOffset()).isEqualTo(-1L);
+      }
+    }
```
I think this is great validation we should add, but I think in the Spark tests we should use the Spark APIs or Spark SQL to perform the write and then we run the validation to make sure the bloom filters exist. That should help catch the issue why we're not seeing bloom filters being written for nested types when writing via Spark (but going through `FileAppender` writes masks that)

amogh-jahagirdar replied:
Also double check style (if you run `./gradlew spotlessApply` it should fix all the current issues)

hussein-awala replied:
> I think in the Spark tests we should use the Spark APIs

I totally agree, I will update it.

hussein-awala commented:
@amogh-jahagirdar I tried to use spark to write the data (check https://github.com/apache/iceberg/pull/9902/commits/25815479628015143551c2379be5608e2dd09bd7), but I had an issue with the used metastore (Hive thrift):
```
Serialization stack:
	- object not serializable (class: org.apache.hadoop.fs.Path, value: file:/var/folders/1q/drsr0xqn0mzf03hhhf1z67_40000gn/T/hive12824930955088255332/test/.hive-staging_hive_2024-03-09_23-55-13_177_3570944359873831448-1/-ext-10000)
	- field (class: org.apache.hadoop.hive.ql.plan.FileSinkDesc, name: dirName, type: class org.apache.hadoop.fs.Path)
	- object (class org.apache.hadoop.hive.ql.plan.FileSinkDesc, org.apache.hadoop.hive.ql.plan.FileSinkDesc@826a408)
	- field (class: org.apache.spark.sql.hive.execution.HiveFileFormat$$anon$1, name: fileSinkConfSer$1, type: class org.apache.hadoop.hive.ql.plan.FileSinkDesc)
	- object (class org.apache.spark.sql.hive.execution.HiveFileFormat$$anon$1, org.apache.spark.sql.hive.execution.HiveFileFormat$$anon$1@2d533ec6)
	- field (class: org.apache.spark.sql.execution.datasources.WriteJobDescription, name: outputWriterFactory, type: class org.apache.spark.sql.execution.datasources.OutputWriterFactory)
	- object (class org.apache.spark.sql.execution.datasources.WriteJobDescription, org.apache.spark.sql.execution.datasources.WriteJobDescription@260edd7e)
	- element of array (index: 0)
	- array (class [Ljava.lang.Object;, size 4)
	- field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
	- object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.apache.spark.sql.execution.datasources.WriteFilesExec, functionalInterfaceMethod=scala/Function1.apply:(Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/apache/spark/sql/execution/datasources/WriteFilesExec.$anonfun$doExecuteWrite$1:(Lorg/apache/spark/sql/execution/datasources/WriteJobDescription;Ljava/lang/String;Lorg/apache/spark/internal/io/FileCommitProtocol;Lscala/Option;Lscala/collection/Iterator;)Lscala/collection/Iterator;, instantiatedMethodType=(Lscala/collection/Iterator;)Lscala/collection/Iterator;, numCaptured=4])
	- writeReplace data (class: java.lang.invoke.SerializedLambda)
	- object (class org.apache.spark.sql.execution.datasources.WriteFilesExec$$Lambda$2347/0x000000e001f3c5f8, org.apache.spark.sql.execution.datasources.WriteFilesExec$$Lambda$2347/0x000000e001f3c5f8@34bb023a)
	- element of array (index: 0)
	- array (class [Ljava.lang.Object;, size 1)
	- field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
	- object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.apache.spark.rdd.RDD, functionalInterfaceMethod=scala/Function3.apply:(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/apache/spark/rdd/RDD.$anonfun$mapPartitionsInternal$2$adapted:(Lscala/Function1;Lorg/apache/spark/TaskContext;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, instantiatedMethodType=(Lorg/apache/spark/TaskContext;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, numCaptured=1])
	- writeReplace data (class: java.lang.invoke.SerializedLambda)
	- object (class org.apache.spark.rdd.RDD$$Lambda$2349/0x000000e001f3cca8, org.apache.spark.rdd.RDD$$Lambda$2349/0x000000e001f3cca8@68ee800f)
	- field (class: org.apache.spark.rdd.MapPartitionsRDD, name: f, type: interface scala.Function3)
	- object (class org.apache.spark.rdd.MapPartitionsRDD, MapPartitionsRDD[3] at create at TestSparkReaderWithBloomFilter.java:210)
	- field (class: scala.Tuple2, name: _1, type: class java.lang.Object)
	- object (class scala.Tuple2, (MapPartitionsRDD[3] at create at TestSparkReaderWithBloomFilter.java:210,org.apache.spark.sql.execution.datasources.FileFormatWriter$$$Lambda$2354/0x000000e001f44d28@6ff879e5))
``` 
Do you have any idea?

I tried to remove the metastore and use `IcebergSparkSessionExtensions` as SQL extension, but the class is not available in the module, I can check how to fix it if you prefer this solution since it's the one explained in Iceberg documentation for Spark users.

hussein-awala commented:
I just fixed it, and I updated the tests for the other versions to make them similar to 3.5.

It looks like bloom filters are written by Spark, I will re-test my jobs to understand what is the issue.

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderWithBloomFilter.java:
```diff
@@ -174,170 +156,62 @@ public static Object[][] parameters() {
-  }
+    createTableWriter.tableProperty(PARQUET_ROW_GROUP_SIZE_BYTES, "100");
 
-  private DataFile writeDataFile(OutputFile out, StructLike partition, List<Record> rows)
```
this seems like too many changes just to add a single test. This makes it quite difficult to review the diffset

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderWithBloomFilter.java:
```diff
@@ -18,56 +18,43 @@
 import static org.apache.iceberg.TableProperties.PARQUET_ROW_GROUP_SIZE_BYTES;
-import static org.apache.iceberg.TableProperties.PARQUET_ROW_GROUP_SIZE_BYTES_DEFAULT;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
please use AssertJ assertions (which is already available in the live just above this)

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderWithBloomFilter.java:
```diff
@@ -116,54 +114,38 @@ public class TestSparkReaderWithBloomFilter {
   private static final String BINARY_PREFIX = "BINARY测试_";
 
-  @TempDir private Path temp;
+  @TempDir private static Path temp;
```
why is this change necessary?

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderWithBloomFilter.java:
```diff
@@ -116,54 +114,38 @@ public class TestSparkReaderWithBloomFilter {
-                  new BigDecimal(String.valueOf(88.88)),
-                  "id_fixed_decimal",
-                  new BigDecimal(String.valueOf(99.99)))));
+      Row row =
```
why are all these changes necessary?

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderWithBloomFilter.java:
```diff
@@ -95,18 +80,31 @@ public class TestSparkReaderWithBloomFilter {
-          Types.NestedField.optional(10, "id_fixed_decimal", Types.DecimalType.of(31, 2)));
+  public static final StructType schema =
+      new StructType(
+          new StructField[] {
```
why are these changes necessary?

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderWithBloomFilter.java:
```diff
@@ -351,12 +225,13 @@ public void testReadWithFilter() {
                     + " AND id_string = 'BINARY测试_30' AND id_boolean = true AND id_date = '2021-09-05'"
-                    + " AND id_int_decimal = 77.77 AND id_long_decimal = 88.88 AND id_fixed_decimal = 99.99");
+                    + " AND id_int_decimal = 77.77 AND id_long_decimal = 88.88 AND id_fixed_decimal = 99.99"
+                    + " AND id_nested.nested_id = 30");
```
this change doesn't seem related to the PR title

nastra reviewed (changes requested):
there are a bunch of changes that seem unrelated to what's being proposed to be done (aka adding a check to see if blook filters are added)

hussein-awala commented:
> there are a bunch of changes that seem unrelated to what's being proposed to be done (aka adding a check to see if blook filters are added)

Thanks @nastra for the review. The data was written using an Avro writer, so if there is a problem with the Spark writer, the bug will not be detectable. I refecatored the tests as @amogh-jahagirdar suggested (https://github.com/apache/iceberg/pull/9902#discussion_r1518451041) and I completely agree with him on this point.

nastra commented:
> > there are a bunch of changes that seem unrelated to what's being proposed to be done (aka adding a check to see if blook filters are added)
> 
> Thanks @nastra for the review. The data was written using an Avro writer, so if there is a problem with the Spark writer, the bug will not be detectable. I refecatored the tests as @amogh-jahagirdar suggested ([#9902 (comment)](https://github.com/apache/iceberg/pull/9902#discussion_r1518451041)) and I completely agree with him on this point.

In that case I would probably just add a new test class where reading and writing is done through Spark. I think the purpose of `TestSparkReaderWithBloomFilter` was to actually only read through Spark

hussein-awala commented:
> In that case I would probably just add a new test class where reading and writing is done through Spark. I think the purpose of TestSparkReaderWithBloomFilter was to actually only read through Spark

Yes I can do that, but some logics will be duplicated (data writer for avro and spark, random data generator, ...), knowing that replacing avro writer with spark df write reduced the code size (`+435 −742` despite the addition of a new test), and simplified it. If you check the logic of the existing test, it didn't change, I just updated the extraction of row values to do the assertion. WDYT?
If you still consider it an unnecessary change, I can create a new class for the new test.

nastra commented:
> > In that case I would probably just add a new test class where reading and writing is done through Spark. I think the purpose of TestSparkReaderWithBloomFilter was to actually only read through Spark
> 
> Yes I can do that, but some logics will be duplicated (data writer for avro and spark, random data generator, ...), knowing that replacing avro writer with spark df write reduced the code size (`+435 −742` despite the addition of a new test), and simplified it. If you check the logic of the existing test, it didn't change, I just updated the extraction of row values to do the assertion. WDYT? If you still consider it an unnecessary change, I can create a new class for the new test.

I'm still +1 on just adding a separate test class and not change the existing test (since that effectively changes the write path).
